sql/mysqld.cc

```diff
@@ -2092,33 +2092,17 @@ static void clean_up(bool print_message)
 */
 static void wait_for_signal_thread_to_end()
 {
-  uint i, n_waits= DBUG_IF("force_sighup_processing_timeout") ? 5 : 100;
-  int err= 0;
+#ifndef _WIN32
   /*
     Wait up to 10 seconds for signal thread to die. We use this mainly to
     avoid getting warnings that my_thread_end has not been called
   */
-  for (i= 0 ; i < n_waits && signal_thread_in_use; i++)
-  {
-    err= pthread_kill(signal_thread, MYSQL_KILL_SIGNAL);
-    if (err)
-      break;
-    my_sleep(100000); // Give it time to die, .1s per iteration
-  }
-
-  if (err && err != ESRCH)
+  for (uint i= 0 ; i < 100 && signal_thread_in_use; i++)
```

Comment:
https://buildbot.mariadb.org/#/builders/572/builds/11911 is failing, this is testing this code.

The current code in 11.6 is slightly different https://github.com/MariaDB/server/blob/68e369e3a97940303173b124fdfb7c626effbe6b/sql/mysqld.cc#L2095

It retains kill() instead of pthread_kill(), like here, but also has some warnings, if signal thread did not shutdown in time, and pthread_join(), that was removed here

Reply:
yeah, thx for point this out, seems like I was a bit confused by the amounts of changes in https://github.com/MariaDB/server/blob/2f5174e556cd247133aa14d7b37372ae49fe83c5/sql/mysqld.cc#L2074-L2088 that did not left anything from this func. I will fix it.

however, speaking of failures its interesting that its actually failing for Debian and Ubuntu in a segfault assert in https://github.com/MariaDB/server/blob/10.11/mysql-test/suite/rpl/t/rpl_shutdown_sighup.test (I am not sure which commit it uses in https://buildbot.mariadb.org/#/builders/369/builds/23487/steps/7/logs/stdio : line 3404).

Anyway it seems like it segfaults in signal_hand() that is a different function, and i did expect it not to call a reload:
```
[Current thread is 1 (Thread 0x7f5f147a5700 (LWP 104156))]
#0  0x00007f5f227fcf68 in pthread_kill () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00005625f2a46126 in my_write_core (sig=11) at /home/buildbot/amd64-ubuntu-2004-debug/build/mysys/stacktrace.c:424
#2  0x00005625f21b8c3e in handle_fatal_signal (sig=11) at /home/buildbot/amd64-ubuntu-2004-debug/build/sql/signal_handler.cc:358
#3  <signal handler called>
#4  0x00005625f1d50b5d in hash_filo::clear (this=0x0, locked=false) at /home/buildbot/amd64-ubuntu-2004-debug/build/sql/hash_filo.h:87
#5  0x00005625f21d6cb2 in hostname_cache_refresh () at /home/buildbot/amd64-ubuntu-2004-debug/build/sql/hostname.cc:134
#6  0x00005625f20157a2 in reload_acl_and_cache (thd=0x0, options=2147499823, tables=0x0, write_to_binlog=0x7f5f147a4c10) at /home/buildbot/amd64-ubuntu-2004-debug/build/sql/sql_reload.cc:379
[-------------------]
#7  0x00005625f1c84d21 in signal_hand () at /home/buildbot/amd64-ubuntu-2004-debug/build/sql/mysqld.cc:3277
#8  0x00005625f24aa749 in pfs_spawn_thread (arg=0x5625f6888268) at /home/buildbot/amd64-ubuntu-2004-debug/build/storage/perfschema/pfs.cc:2201
#9  0x00007f5f227f4609 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#10 0x00007f5f223c5353 in clone () from /lib/x86_64-linux-gnu/libc.so.6
```

I will look into that.

Reply:
The amount of changes to this function is has to do with a different bug,  MDEV-30260: Slave crashed:reload_acl_and_cache during shutdown , which they fixed after my changes that you have ported.
This is 952ab9a59659b28f10ce2770e17249e0ff5b0681 

It is a rare condition, where SIGHUP signal handler (which does "reloads" in signal handler thread) coincidences with shutdown.

Will it still be working OK, if you just copy that function verbatim from 11.6 (I think 11.6 should work ok on macOS and FreeBSD etc, I think)?

Also, it makes sense to base the patch tor the earliest affected and still supported branch, which is 10.5. It will be merged to all subsequent versions anyway.

Reply:
Got you @vaintroub. Surprisingly I have eliminated pthread_join in a previous version. Reverted needed pieces back. Now it must be ok. I have ported work to 10.5 at https://github.com/MariaDB/server/pull/3469, the code there is a bit different for termiation_fd commit so had to keep it separate.

Also, any ideas why this may fail? Does it fail sometimes or probably I missed something again? Didn't look at logs yet:

```
Completed: All 6112 tests were successful.

Errors/warnings were found in logfiles during server shutdown after running the
following sequence(s) of tests:
    binlog_encryption.rpl_parallel_missed_error_handling

298 tests were skipped, 213 by the test itself.

mysql-test-run: *** ERROR: There where errors/warnings in server logs after running test cases.
```

Reply:
```
***Warnings generated in error logs during shutdown after running tests: binlog_encryption.rpl_parallel_missed_error_handling
Warning: Memory not freed: -2072
```

2072 is an interesting number. Close to 2KB.

Reply:
2nd attempt has different failure. https://buildbot.mariadb.org/#/builders/640/builds/4586/steps/7/logs/stdio

```
Program terminated with signal SIGABRT, Aborted.
#0  0x00007f13c8737f44 in pthread_kill () from /lib/x86_64-linux-gnu/libpthread.so.0
[Current thread is 1 (Thread 0x7f13c7eb3980 (LWP 230258))]
#0  0x00007f13c8737f44 in pthread_kill () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x000055f48e7c92ca in handle_fatal_signal ()
#2  0x000055f48daf333b in SignalHandler(int) ()
#3  <signal handler called>
#4  0x00007f13c82219af in poll () from /lib/x86_64-linux-gnu/libc.so.6
#5  0x000055f48dac3ff3 in poll ()
#6  0x000055f48db1bce5 in handle_connections_sockets() ()
#7  0x000055f48db0bac2 in mysqld_main(int, char**) ()
#8  0x00007f13c8155d0a in __libc_start_main () from /lib/x86_64-linux-gnu/libc.so.6
#9  0x000055f48da7945a in _start ()
```

during atomic.alter_table.

@vaintroub how do you think, its my changes or the msan ci tests are flaky?

Reply:
Does not look like your changes. That test ran into timeout, and was terminated with SIGABRT, which mtr does to produce coredump, to analyze a potential hang with gdb . but there is nothing interesting in the crash , there is not even shutdown or anything. Maybe it just took too long.

Reply:
yes, this test [regularly runs into timeout,](https://buildbot.mariadb.net/ci/reports/cross_reference#branch=&revision=&platform=&fail_name=atomic.alter_table&fail_variant=&fail_info_full=&typ=&info=&dt=&limit=100&fail_info_short=) and very often on "msan" builders

Reply:
should I re-execute CI until it succeeds?

Reply:
I do not think it is necessary